blockchain/validate.go

```diff
@@ -767,6 +756,28 @@ func (b *BlockChain) checkBlockContext(block *btcutil.Block, prevNode *blockNode
 		return err
 	}
 
+	// Query for the Version Bits state for the segwit soft-fork
+	// deployment. If segwit is active, we'll switch over to enforcing all
+	// the new rules.
+	segwitState, err := b.deploymentState(prevNode, chaincfg.DeploymentSegwit)
```

Comment:
This whole section should be moved under the `if !fastAdd` branch.  I would also recommend moving it after the `BIP0034Height` check because it's somewhat expensive to validate the witness commitment due to needing to serialize and hash all of the transactions again with the segwit encoding, so it's better to do it after the cheaper checks.